File: internal/service/sketch.go
package service

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// MakeSketch converts an image to a pencil-sketch style PNG.
// sigma controls line softness (0.5 - 10).
func MakeSketch(tmpDir, src string, sigma float64) (string, error) {
	if sigma <= 0 {
		sigma = 3.0
	}
	if sigma > 10 {
		sigma = 10.0
	}
	dest := filepath.Join(tmpDir, "sketch.png")
	// Convert input into PNG first (handles BMP/WebP etc.)
	pngPath := filepath.Join(tmpDir, "sketch_input.png")
	if out, err := RunCmd(PythonPath(), ScriptPath("vectorize.py"), "topng", src, pngPath); err != nil {
		return "", fmt.Errorf("图片读取失败: %s", strings.TrimSpace(out))
	}
	sigmaStr := strconv.FormatFloat(sigma, 'f', -1, 64)
	out, err := RunCmd(PythonPath(), ScriptPath("sketch.py"), "go", pngPath, sigmaStr, dest)
	if err != nil {
		return "", fmt.Errorf("素描生成失败: %s", strings.TrimSpace(out))
	}
	if _, err := os.Stat(dest); err != nil {
		return "", fmt.Errorf("素描生成失败")
	}
	return dest, nil
}

File: src/service/sketch.rs
use crate::util::{run_cmd, python_path, script_path};
use std::path::PathBuf;

pub fn make_sketch(tmp_dir: &str, src: &str, sigma: f64) -> Result<String, String> {
    let sigma = if sigma <= 0.0 { 3.0 } else if sigma > 10.0 { 10.0 } else { sigma };
    let dest = PathBuf::from(tmp_dir).join("sketch.png");
    let png_path = PathBuf::from(tmp_dir).join("sketch_input.png");

    // Convert input to PNG first
    let result = run_cmd(python_path(), &[
        &script_path("vectorize.py"),
        "topng",
        src,
        png_path.to_str().unwrap(),
    ]);
    if let Some(ref err) = result.error {
        return Err(format!("图片读取失败: {}", err));
    }

    let sigma_str = sigma.to_string();
    let result = run_cmd(python_path(), &[
        &script_path("sketch.py"),
        "go",
        png_path.to_str().unwrap(),
        &sigma_str,
        dest.to_str().unwrap(),
    ]);
    if let Some(ref err) = result.error {
        return Err(format!("素描生成失败: {}", err));
    }
    if !dest.exists() {
        return Err("素描生成失败".to_string());
    }
    Ok(dest.to_string_lossy().to_string())
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_sigma_clamping() {
        assert!((3.0_f64 - 3.0_f64).abs() < 0.001);
        assert!((10.0_f64 - 10.0_f64).abs() < 0.001);
        assert!((5.0_f64 - 5.0_f64).abs() < 0.001);
    }
}
